### ISSUE TITLE
[jaeger] Conditionally deploy jaeger agent service and

### ISSUE THREAD

**Issue by mehta-ankit** [jaegertracing/helm-charts#249]
I was playing with jaeger agents deployment as a daemonset and few thoughts came to my mind:

- We currently deploy agent service when agents are enabled: https://github.com/jaegertracing/helm-charts/blob/main/charts/jaeger/templates/agent-svc.yaml
Is this needed ? If i am deploying agents as a daemonset i would use the Node IP and set it as JAEGER_AGENT_HOST env variable on the app and use hostPort on the agents to communicate with them. Is the service created so a servicemonitor can be used to get the agent metrics ??
that can be done using a podMonitor, no ?

- In the daemonset spec, hostPorts are created for all ports (thrift, compact , binary and samplingport): https://github.com/jaegertracing/helm-charts/blob/main/charts/jaeger/templates/agent-ds.yaml#L72
This should be conditional, so a user can only create hostPorts that they need. For example based on the client application they are using: https://www.jaegertracing.io/docs/1.22/client-features/, they would need 2 of these ports at max. Many clusters have security policies against hostPorts and if someone has to create a PSP to go around them, its better they only do it for ports that are needed for their use case (otherwise another pod using the same port cannot be scheduled on the specific node).

@naseemkullah @pavelnikolov let me know your thoughts ??

**Comment by naseemkullah** [comment]
I agree with both those statements, though for the second if possible keep as is and allow user to override the array of allowed ports, in yaml since you cant concat arrays this should be good

**Comment by mehta-ankit** [comment]
`for the second if possible keep as is and allow user to override the array of ports, in yaml since you cant concat arrays`
Can you clarify with an example here please. I am not sure i understand this comment :)
In my head i wanted this section to be conditional
```
{{- if .Values.agent.daemonset.useHostPort }} 
hostPort: {{ .Values.agent.service.compactPort }}
{{- end }}
```

**Comment by naseemkullah** [comment]
> `for the second if possible keep as is and allow user to override the array of ports, in yaml since you cant concat arrays`
> Can you clarify with an example here please. I am not sure i understand this comment :)
> In my head i wanted this section to be conditional, as currently based on the boolean value of `useHostPort` they all show up or none.
> 
> ```
> {{- if .Values.agent.daemonset.useHostPort }} 
> hostPort: {{ .Values.agent.service.compactPort }}
> {{- end }}
> ```
> 
> Also i saw if useHostNetwork is true which sets hostNetwork to true on the pod, all hostPorts automatically show up even when this: `.Values.agent.daemonset.useHostPort` was `false`:. I am assuming its some K8s thing 🤷🏽
> 
> ```
>  {{- if .Values.agent.useHostNetwork }}
>  hostNetwork: true
>  {{- end }}
> ```

In my case I use the binary port (6832) because our services are in node.js, so basically i think we leave the ports on by default to not break things, but we should add some logic in values file to control https://github.com/jaegertracing/helm-charts/blob/c08ca10c5cf4f2d4bfaf64772ad61f472c00e2f1/charts/jaeger/templates/agent-ds.yaml#L71-L95 such that a user, can set these https://github.com/jaegertracing/helm-charts/blob/c08ca10c5cf4f2d4bfaf64772ad61f472c00e2f1/charts/jaeger/values.yaml#L214-L220 to null to disable them

my comment about a ports array does not actually apply here, my bad
